Generate and integrate introspection query results

wyattjoh opened the pull request:
## What does this PR do?

- Added new WEBPACK env var which is enabled during yarn build scripts
- Defered redis starting until listen is called
- Moved pubsub to a factory pattern init
- Async/Await'ed the routes
- Moved pubsub handle for routes into middleware
- Adjusted redis cache and job processors to have lazy connection
starting
- Disabled mongo from auto-connecting on require
- Adjusted package redis clients to act as factory singletons instead
- Generate a `introspection.json` as part of our build process.
- Auto update `introspection.json` when using `yarn dev-start`.
- Port client code to use the `IntrospectionFragmentMatcher`.

## How do I test this PR?

1. Stop mongo + redis docker containers.
2. Create a file `test_772.js` with the contents:

```js
require('./app');
```

3. Run `WEBPACK=true NODE_ENV=test node test_772.js`
4. See that the script starts/stops without issue.

wyattjoh committed:
Adjust redis to not start during webpack build

- Added new WEBPACK env var which is enabled during yarn build scripts
- Defered redis starting until listen is called
- Moved pubsub to a factory pattern init
- Async/Await'ed the routes
- Moved pubsub handle for routes into middleware
- Adjusted redis cache and job processors to have lazy connection
starting
- Disabled mongo from auto-connecting on require
- Adjusted package redis clients to act as factory singletons instead

kgardnr committed:
Merge branch 'master' into next

kgardnr committed:
Merge branch 'master' into next

kgardnr committed:
Merge branch 'master' into next

kgardnr committed:
Merge branch 'master' into next

kgardnr committed:
Merge branch 'master' into next

wyattjoh committed:
Merge branch 'master' into next

alinhle commented on services/scraper.js:
```diff
@@ -12,11 +12,20 @@ const metascraper = require('metascraper');
-  task: new kue.Task({
-    name: 'scraper'
-  }),
+  _task: null,
```
I think `_task` should be kept internal like:

```js
let taskInstance = null;

const scraper = {
  // ...
  get task() {
    if (taskInstance) {
      return taskInstance;
    }

    taskInstance = new kue.Task({
      name: 'scraper'
    });

    return taskInstance;
  },
  // ...
};
```

alinhle commented on services/mailer.js:
```diff
@@ -75,9 +75,18 @@ const mailer = module.exports = {
-  task: new kue.Task({
-    name: 'mailer'
-  }),
+  _task: null,
```
I think _task should be kept internal. See below.

alinhle commented on services/mongoose.js:
```diff
@@ -40,14 +41,23 @@ if (enabled('talk:db')) {
+if (WEBPACK) {
+
+  console.warn('Not connecting to mongodb during webpack build');
+  mongoose.disconnect();
```
So mongoose automatically connects and we have to disconnect manually when in Webpack build?

wyattjoh replied:
Not exactly. We don't call connect, but because we include it, it will hold the socket ready, preventing node from exiting. Calling disconnect here just ensures that the application can quit correctly.

alinhle replied:
Would be great if you add that as a comment :-)

alinhle commented on services/passport.js:
```diff
@@ -9,12 +9,21 @@ const LocalStrategy = require('passport-local').Strategy;
-const client = createClient();
+const {createClient} = require('./redis');
+let _client = null;
+const getClient = () => {
```
Why not export a `getClient` from `./redis.js` and reuse it here?

```js
// redis.js
let redisInstance = null;
module.exports = {
  // ...
  getClient() {
    if (redisInstance) {
      return redisInstance;
    }
    redisInstance = createClient();
    return redisInstance;
  },
  // ...
}
```

alinhle commented on services/users.js:
```diff
@@ -32,6 +29,19 @@ const PASSWORD_RESET_JWT_SUBJECT = 'password_reset';
+// Create a redis client to use for authentication.
+const {createClient} = require('./redis');
+let _client = null;
+const getClient = () => {
```
See above

alinhle commented on services/pubsub.js:
```diff
@@ -2,4 +2,13 @@ const {RedisPubSub} = require('graphql-redis-subscriptions');
 
-module.exports = new RedisPubSub({connection: connectionOptions});
+let pubsub = null;
+module.exports = () => {
```
Maybe exporting an object with a descriptive function is more clear:

```js
// pubsub.js
module.exports = {
  getInstance() {
    if (pubsub) {
      return pubsub;
     }
    pubsub = new RedisPubSub({connection: connectionOptions});
  },
}
```

So it can be consumed like this:

```js
const pubsub = require('pubsub');
const pubsubInstance = pubsub.getInstance();
```

Instead of

```js
const pubsub = require('pubsub');
const pubsubInstance = pubsub();
```
The line `pubsub = pubsub()` as used in this PR was not immediately clear to me what it meant.

alinhle commented on services/kue.js:
```diff
@@ -8,18 +8,24 @@ const kue = module.exports.kue = require('kue');
-    createClientFactory: () => redis.createClient()
-  }
-});
+let Queue = module.exports.queue;
```
Doesn't this equal to `let Queue = undefined`? If so rather set it to `let Queue = null;` and put a comment that the instance is lazily created and assigned.

wyattjoh committed:
Merge branch 'master' into next

wyattjoh committed:
Suggestions for refactoring

wyattjoh committed:
Fixes to impl

wyattjoh committed:
removed unused var

wyattjoh committed:
Refined task singleton to factory

wyattjoh committed:
Added introspection results to build pipeline

wyattjoh committed:
Added log about result

alinhle commented on services/redis.js:
```diff
@@ -29,21 +29,36 @@ const connectionOptions = {
-      if (err) {
-        console.error('Can\'t ping the redis server!');
+  createClient,
+  createClientFactory: () => {
```
I saw you use a Factory Pattern to create client instances to redis. Wouldn’t we just want to have one client instance that connects to redis, and as such a singleton pattern would be more appropiate (or written in the factory pattern style but always returns the same instance)?

alinhle committed:
More integration with the introspection query result

- Auto generate introspection on any change when using  `dev-start`
- Port client to use introspection matcher

alinhle committed:
Add comment

alinhle committed:
Merge branch 'master' into next

kgardnr committed:
Merge branch 'master' into next
